components/server/ee/src/workspace/gitpod-server-impl.ts

```diff
@@ -1827,6 +1830,54 @@ export class GitpodServerEEImpl extends GitpodServerImpl {
         return subscription;
     }
 
+    protected async ensureIsUsageBasedFeatureFlagEnabled(user: User): Promise<void> {
+        const teams = await this.teamDB.findTeamsByUser(user.id);
+        const isUsageBasedBillingEnabled = await getExperimentsClient().getValueAsync(
+            "isUsageBasedBillingEnabled",
+            false,
+            {
+                identifier: user.id,
+                custom: {
+                    team_ids: teams.map((t) => t.id).join(","),
+                    team_names: teams.map((t) => t.name).join(","),
+                },
+            },
+        );
+        if (!isUsageBasedBillingEnabled) {
+            throw new ResponseError(ErrorCodes.PERMISSION_DENIED, "not allowed");
+        }
+    }
+
+    async getStripePublishableKey(ctx: TraceContext): Promise<string | undefined> {
+        const user = this.checkAndBlockUser("getStripePublishableKey");
+        await this.ensureIsUsageBasedFeatureFlagEnabled(user);
+        return this.config.stripeSettings?.publishableKey;
+    }
+
+    async getStripeSetupIntentClientSecret(ctx: TraceContext): Promise<string | undefined> {
+        const user = this.checkAndBlockUser("getStripeSetupIntentClientSecret");
+        await this.ensureIsUsageBasedFeatureFlagEnabled(user);
+        const setupIntent = await this.stripeService.createSetupIntent();
```

Comment:
The service implementation doesn't catch errors, and neither does this handler. This means that 3rd party errors (from a library) would be propagated to the client. Could you please throw non-detailed error to the client and log the underlying error such that we don't accidentally leak details?

The same also goes for `getTeamStripeCustomerId` and `subscribeTeamToStripe`

Related thread: https://github.com/gitpod-io/gitpod/pull/10378/files#r887842876

Reply:
I left this as is, because forwarding errors makes debugging a bit easier while this is still behind a feature flag (and we have the same problem in many other server API methods, so a more holistic fix would be a good idea).

However, it's pretty easy to add a try/catch to the Stripe methods I've added, so I'll do that now, hoping that we can then merge this and make progress.

Reply:
Resolved in [[server] Sanitize Stripe error messages sent by server](https://github.com/gitpod-io/gitpod/pull/10378/commits/d31c9f0af53fdd4f088fd028f3506b1905be616b)